lib/web_ui/lib/src/engine/html/shaders/normalized_gradient.dart

```diff
@@ -0,0 +1,161 @@
+      thresholds[thresholdIndex++] = 0.0;
+    }
+    for (ui.Color c in colors) {
+      bias[targetIndex++] = c.red / 255.0;
+      bias[targetIndex++] = c.green / 255.0;
+      bias[targetIndex++] = c.blue / 255.0;
+      bias[targetIndex++] = c.alpha / 255.0;
+    }
+    for (double stop in stops) {
+      thresholds[thresholdIndex++] = stop;
+    }
+    if (addLast) {
+      ui.Color c = colors.last;
+      bias[targetIndex++] = c.red / 255.0;
+      bias[targetIndex++] = c.green / 255.0;
+      bias[targetIndex++] = c.blue / 255.0;
+      bias[targetIndex++] = c.alpha / 255.0;
+      thresholds[thresholdIndex++] = 1.0;
+    }
+    // Now that we have bias for each color stop, we can compute scale based
+    // on delta between colors.
+    int lastColorIndex = 4 * (normalizedCount - 1);
+    for (int i = 0; i < lastColorIndex; i++) {
+      int thresholdIndex = i >> 2;
+      scale[i] = (bias[i + 4] - bias[i]) /
+          (thresholds[thresholdIndex + 1] - thresholds[thresholdIndex]);
+    }
+    scale[lastColorIndex] = 0.0;
+    scale[lastColorIndex + 1] = 0.0;
+    scale[lastColorIndex + 2] = 0.0;
+    scale[lastColorIndex + 3] = 0.0;
+    // Compute bias = colorAtStop - stopValue * (scale).
+    for (int i = 0; i < normalizedCount; i++) {
+      double t = thresholds[i];
+      int colorIndex = i * 4;
+      bias[colorIndex] -= t * scale[colorIndex];
+      bias[colorIndex + 1] -= t * scale[colorIndex + 1];
+      bias[colorIndex + 2] -= t * scale[colorIndex + 2];
+      bias[colorIndex + 3] -= t * scale[colorIndex + 3];
+    }
+    return NormalizedGradient._(normalizedCount, thresholds, scale, bias);
+  }
+
+  /// Sets uniforms for threshold, bias and scale for program.
+  void setupUniforms(_GlContext gl, _GlProgram glProgram) {
+    for (int i = 0; i < thresholdCount; i++) {
+      Object biasId = gl.getUniformLocation(glProgram.program, 'bias_$i');
+      gl.setUniform4f(biasId, _bias[i * 4], _bias[i * 4 + 1], _bias[i * 4 + 2], _bias[i * 4 + 3]);
+      Object scaleId = gl.getUniformLocation(glProgram.program, 'scale_$i');
+      gl.setUniform4f(scaleId, _scale[i * 4], _scale[i * 4 + 1], _scale[i * 4 + 2], _scale[i * 4 + 3]);
+    }
+    for (int i = 0; i < _thresholds.length; i += 4) {
+      Object thresId = gl.getUniformLocation(glProgram.program, 'threshold_${i ~/ 4}');
+      gl.setUniform4f(thresId, _thresholds[i], _thresholds[i + 1], _thresholds[i + 2], _thresholds[i + 3]);
+    }
+  }
+
+  /// Returns bias component at index.
+  double biasAt(int index) => _bias[index];
+
+  /// Returns scale component at index.
+  double scaleAt(int index) => _scale[index];
+
+  /// Returns threshold at index.
+  double thresholdAt(int index) => _thresholds[index];
+}
+
+/// Writes code to search for probe value in source data and set
+/// bias and scale to be used for computation.
+///
+/// Source data for thresholds is provided using ceil(count/4) packed vec4
+/// uniforms.
+/// bias and scale data are vec4 uniforms that hold color data.
```

Comment:
nit: capitalize "Bias" and separate the paragraph from the previous paragraph by one blank line.

Reply:
Done